src/evaluators/DoWhileStatement.js

```diff
@@ -33,34 +34,67 @@ export default function(
 
-    // b. If LoopContinues(stmt, labelSet) is false, return Completion(UpdateEmpty(stmt, V)).
-    if (LoopContinues(realm, stmt, labelSet) === false) {
-      invariant(stmt instanceof AbruptCompletion);
-      // ECMA262 13.1.7
-      if (stmt instanceof BreakCompletion) {
-        if (!stmt.target) return (UpdateEmpty(realm, stmt, V): any).value;
+      // b. If LoopContinues(stmt, labelSet) is false, return Completion(UpdateEmpty(stmt, V)).
+      if (LoopContinues(realm, stmt, labelSet) === false) {
+        invariant(stmt instanceof AbruptCompletion);
+        // ECMA262 13.1.7
+        if (stmt instanceof BreakCompletion) {
+          if (!stmt.target) return (UpdateEmpty(realm, stmt, V): any).value;
+        }
+        throw UpdateEmpty(realm, stmt, V);
       }
-      throw UpdateEmpty(realm, stmt, V);
-    }
 
-    // c. If stmt.[[Value]] is not empty, let V be stmt.[[Value]].
-    let resultValue = InternalGetResultValue(realm, stmt);
-    if (!(resultValue instanceof EmptyValue)) V = resultValue;
+      // c. If stmt.[[Value]] is not empty, let V be stmt.[[Value]].
+      let resultValue = InternalGetResultValue(realm, stmt);
+      if (!(resultValue instanceof EmptyValue)) V = resultValue;
 
-    // d. Let exprRef be the result of evaluating Expression.
-    let exprRef = env.evaluate(test, strictCode);
+      // d. Let exprRef be the result of evaluating Expression.
+      let exprRef = env.evaluate(test, strictCode);
 
-    // e. Let exprValue be ? GetValue(exprRef).
-    let exprValue = Environment.GetValue(realm, exprRef);
+      // e. Let exprValue be ? GetValue(exprRef).
+      let exprValue = Environment.GetConditionValue(realm, exprRef);
+
+      // f. If ToBoolean(exprValue) is false, return NormalCompletion(V).
+      if (To.ToBooleanPartial(realm, exprValue) === false) return V;
+    }
+    invariant(false);
+  });
+  if (resultOrDiagnostic instanceof Value) return resultOrDiagnostic;
 
-    // f. If ToBoolean(exprValue) is false, return NormalCompletion(V).
-    if (To.ToBooleanPartial(realm, exprValue) === false) return V;
+  // If we get here then unrolling the loop did not work, possibly because the value of the loop condition is not known,
+  // so instead try to compute a fixpoint for it
+  let ftest = () => {
+    let exprRef = env.evaluate(test, strictCode);
+    return Environment.GetConditionValue(realm, exprRef);
+  };
+  let fbody = () => env.evaluateCompletion(body, strictCode);
+  let effects = realm.evaluateForFixpointEffects(ftest, fbody);
+  if (effects !== undefined) {
+    let [outsideEffects, insideEffects] = effects;
+    let [rval] = outsideEffects;
+    let [, bodyGenerator] = insideEffects;
+    realm.applyEffects(outsideEffects);
+    let exprRef = env.evaluate(test, strictCode);
+    let exprValue = Environment.GetValue(realm, exprRef);
+    invariant(exprValue instanceof AbstractValue);
+    let cond = bodyGenerator.derive(exprValue.types, exprValue.values, [exprValue], ([n]) => n, {
```

Comment:
Why do you create `cond` wrapping around `exprValue`?

Reply:
Since cond is the loop exit condition it needs to be temporal.

Reply:
I don't see how that's needed. There are no side effects involved here. Is there something concrete that goes wrong when you don't `derive`?

Reply:
I can see how you'd like to get a value with an intrinsic name as that's what `emitDoWhileStatement` assumes, but I don't see a deep reason why we need to give a name to the value.

Reply:
When I don't force evaluation of the expression inside the loop body, it does not actually get emitted anywhere if I do the seemingly obvious thing of making the test condition an argument of the entry created by emitDoWhileStatement. Perhaps we can came back to this when you are in the office again?

Reply:
Okay. Please leave behind a comment / TODO in the code to this end.